src/internal/small_vec.rs

```diff
@@ -0,0 +1,87 @@
+use std::fmt;
+
+#[derive(Clone)]
+pub enum SmallVec<T> {
+    Empty,
+    One([T; 1]),
+    Two([T; 2]),
+    Flexible(Vec<T>),
+}
+
+impl<T> SmallVec<T> {
+    pub fn empty() -> Self {
+        Self::Empty
+    }
+
+    pub fn one(t: T) -> Self {
+        Self::One([t])
+    }
+
+    pub fn as_slice(&self) -> &[T] {
+        match self {
+            Self::Empty => &[],
+            Self::One(v) => v,
+            Self::Two(v) => v,
+            Self::Flexible(v) => v,
+        }
+    }
+
+    pub fn push(&mut self, new: T) {
+        let s = std::mem::replace(self, Self::empty());
+        *self = match s {
+            Self::Empty => Self::One([new]),
+            Self::One([v1]) => Self::Two([v1, new]),
+            Self::Two([v1, v2]) => Self::Flexible(vec![v1, v2, new]),
```

Comment:
Since I believe `Vec` first breaking point in memory allocation is 4 elements, it might be worth checking if adding enum variants for 3 and 4 elements could improve (or not) performances.

Reply:
I did some experiments and adding more variants did not help with the elm benchmark. It makes all Ranges 1/n bigger on the stack for ocasinaly not having to allocatie. It is not surprising that there is a tipping point somewhere. Exactly where optimizes runtime will depend on the size of V, and the size of system cache, an other optimizations. So we should continue to experiment.